src/IECoreScene/CurvesAlgoUpdateEndpointMultiplicity.cpp

```diff
@@ -46,110 +46,74 @@ using namespace Imath;
-
-		// skip the first two values
-		index += 2;
+			for( size_t j = 0; j < numIn; ++j )
+			{
+				out.push_back( in[curveIndex + j] );
+			}
 
-		for( size_t j = 0; j < numSegments + 1; ++j, ++index )
-		{
-			out.push_back( in[index] );
+			// duplicate the end point 
+			for( int j = 0; j < adjust; j++ )
+			{
+				out.push_back( in[curveIndex + numIn - 1] );
+			}
 		}
 
-		// skip the last two values
-		index += 2;
+		curveIndex += numIn;
 	}
 }
 
-template<typename T>
-void expandVarying( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
+// Set the endpoint multiplicities for vertex and varying primvars, based on the given basis
+void setMultiplicity( const IECore::CubicBasisf &cubicBasis, int &vertexMultiplicity, int &varyingMultiplicity )
 {
-	size_t numCurves = curves->numCurves();
-	out.reserve( in.size() + numCurves * 2 );
-
-	size_t index = 0;
-	for( size_t i = 0; i < numCurves; ++i )
+	if( cubicBasis == IECore::CubicBasisf::catmullRom() )
 	{
-		size_t numSegments = CurvesPrimitive::numSegments( targetBasis, curves->periodic(), curves->variableSize( PrimitiveVariable::Vertex, i ) + 4 );
-
-		// duplicate the first segment value
-		out.push_back( in[index] );
-
-		for( size_t j = 1; j < numSegments; ++j, ++index )
-		{
-			out.push_back( in[index] );
-		}
-
-		// duplicate the last segment value (note we've already incremented index for the next curve)
-		out.push_back( in[index-1] );
+		vertexMultiplicity = 2;
+		varyingMultiplicity = 1;
 	}
-}
-
-// remove the replicated first and last values reversing the 'expand' function above
-template<typename T>
-void compressVarying( const std::vector<T> &in, std::vector<T> &out, const CurvesPrimitive *curves, const PrimitiveVariable &primVar, const IECore::CubicBasisf &targetBasis )
-{
-	size_t numCurves = curves->numCurves();
-	out.reserve( in.size() - numCurves * 2 );
-
-	size_t index = 0;
-	for( size_t i = 0; i < numCurves; ++i )
+	else if( cubicBasis == IECore::CubicBasisf::bSpline() )
 	{
-		size_t numSegments = CurvesPrimitive::numSegments( targetBasis, curves->periodic(), curves->variableSize( PrimitiveVariable::Vertex, i ) - 4 );
-
-		// skip the first segment value
-		++index;
-
-		for( size_t j = 0; j < numSegments + 1; ++j, ++index )
-		{
-			out.push_back( in[index] );
-		}
-
-		// skip the last segment value
-		++index;
+		vertexMultiplicity = 3;
+		varyingMultiplicity = 2;
+	}
+	else
+	{
+		vertexMultiplicity = 1;
+		varyingMultiplicity = 1;
 	}
 }
 
 struct DuplicateEndPoints
 {
-	DuplicateEndPoints( const IECore::CubicBasisf &targetBasis, bool expand ) : m_targetBasis( targetBasis ), m_expand( expand )
+	DuplicateEndPoints( const IECore::CubicBasisf &targetBasis, int adjustVertex, int adjustVarying ) : m_targetBasis( targetBasis ), m_adjustVertex( adjustVertex ), m_adjustVarying( adjustVarying )
```

Comment:
similar to my suggest rename above, perhaps `vertexAdjustment` and `varyingAdjustment`?